docs/deep-learning/lora_training.md

```diff
@@ -0,0 +1,143 @@
+# Encrypted fine-tuning
+
+This section explains how to fine-tune neural-network models and large
+language-models on private data. Small models can be be fine-tuned
+using a single-client/single-server setup. For optimal
+latency when fine-tuning larger models (e.g., GPT2 and bigger)
+you should consider distributed computation, with multiple worker nodes performing the
+training on encrypted data.
+
+## Overview
+
+{% hint style="info" %}
+For a tutorial about applying FHE LORA fine-tuning to a small neural network, see [this notebook](../advanced_examples/LoraMLP.ipynb).
+{% endhint %}
+
+Concrete ML supports LORA, a parameter efficient fine-tuning (PEFT) approach, in
+the [hybrid model](../guides/hybrid-models.md) paradigm. LORA adds
+adapters, which contain a low number of fine-tunable weights, to the linear layers
+in an original model.
+
+Concrete ML will outsource the logic of a model's original forward and backward passes
+to one or more remote servers. On the other hand, the forward and backward passes
+over the LORA weights, the loss computation and the weight updates are performed
+by the client side. As the number of LORA weights is low, this does not incur
+significant added computation time for the model training client machine. More than
+99% of a model's weights can be outsourced for large LLMs.
+
+The main benefit of hybrid-model LORA training is outsourcing the computation of the
+linear layers. In LLMs these layers have considerable size and performing inference
+and gradient computations for them requires significant hardware. Using Concrete ML,
+these computations can be securely outsourced.
+
+## Usage
+
+Concrete ML integrates with the [`peft` package](https://huggingface.co/docs/peft/index)
+which adds LORA layer adapters to a model's linear layers. Here are the steps to convert
+a model to hybrid FHE LORA training.
+
+### 1. Apply the `peft` LORA layers
+
+The `LoraConfig` class from the `peft` package contains the various LORA parameters. It
+is possible to specify which layers have LORA adapters through the `target_modules` argument.
+Please refer to the
+[`LoraConfig`](https://huggingface.co/docs/peft/package_reference/lora#peft.LoraConfig)
+documentation for a reference on the various config options.
+
+<!--pytest-codeblocks:skip-->
+
+```python
+# Apply LoRA to the model
+peft_model = get_peft_model(model, peft_config)
+
+lora_config = LoraConfig(
+    r=1, lora_alpha=1, lora_dropout=0.01, target_modules=["fc1", "fc2"], bias="none"
+)
+```
+
+### 2. Convert the LORA model to use custom Concrete ML layers
+
+Concrete ML requires a conversion step for the `peft` model, adding
+FHE compatible layers. In this step the number of gradient accumulation steps
+can also be set. For LORA it is common to accumulate gradients over
+several gradient descent steps before updating weights.
+
+<!--pytest-codeblocks:skip-->
+
+`python`
+gradient_accum_steps = 2
+lora_training = LoraTraining(peft_model, gradient_accum_steps)
+
+````
```

Comment:
I realize we don't handle that gradient accumulation properly. It's not going to be easy to fix we really need to integrate the training loop within the lora training. I will fix this in the second PR.

Currently, users need to handle the `lora_training.run_optimizer(False|True)` every `gradient_accum_steps` iteration. So basically we should probably limit `gradient_accum_steps=1` for now. And not talk about it here.